README.md
# Rough.js

<b>Rough.js</b> is a small (\<9 kB) graphics library that lets you draw in a _sketchy_, _hand-drawn-like_, style.
The library defines primitives to draw lines, curves, arcs, polygons, circles, and ellipses. It also supports drawing [SVG paths](https://developer.mozilla.org/en-US/docs/Web/SVG/Tutorial/Paths).

Rough.js works with both [Canvas](https://developer.mozilla.org/en-US/docs/Web/API/Canvas_API) and [SVG](https://developer.mozilla.org/en-US/docs/Web/SVG).

![Rough.js sample](https://roughjs.com/images/cap_demo.png)

[@RoughLib](https://twitter.com/RoughLib) on Twitter.

## Install

from npm:

```
npm install --save roughjs
```

Or get the latest using unpkg: https://unpkg.com/roughjs@latest/bundled/rough.js


If you are looking for bundled version in different formats, the npm package will have these in the following locations:

CommonJS: `roughjs/bundled/rough.cjs.js`

ESM: `roughjs/bundled/rough.esm.js`

Browser IIFE: `roughjs/bundled/rough.js`


## Usage

![Rough.js rectangle](https://roughjs.com/images/m1.png)

```js
const rc = rough.canvas(document.getElementById('canvas'));
rc.rectangle(10, 10, 200, 200); // x, y, width, height
```

or SVG

```js
const rc = rough.svg(svg);
let node = rc.rectangle(10, 10, 200, 200); // x, y, width, height
svg.appendChild(node);
```

### Lines and Ellipses

![Rough.js rectangle](https://roughjs.com/images/m2.png)

```js
rc.circle(80, 120, 50); // centerX, centerY, diameter
rc.ellipse(300, 100, 150, 80); // centerX, centerY, width, height
rc.line(80, 120, 300, 100); // x1, y1, x2, y2
```

### Filling

![Rough.js rectangle](https://roughjs.com/images/m3.png)

```js
rc.circle(50, 50, 80, { fill: 'red' }); // fill with red hachure
rc.rectangle(120, 15, 80, 80, { fill: 'red' });
rc.circle(50, 150, 80, {
  fill: "rgb(10,150,10)",
  fillWeight: 3 // thicker lines for hachure
});
rc.rectangle(220, 15, 80, 80, {
  fill: 'red',
  hachureAngle: 60, // angle of hachure,
  hachureGap: 8
});
rc.rectangle(120, 105, 80, 80, {
  fill: 'rgba(255,0,200,0.2)',
  fillStyle: 'solid' // solid fill
});
```

Fill styles can be: **hachure**(default), **solid**, **zigzag**, **cross-hatch**, **dots**, **dashed**, or **zigzag-line**

![Rough.js fill examples](https://roughjs.com/images/m14.png)

### Sketching style

![Rough.js rectangle](https://roughjs.com/images/m4.png)

```js
rc.rectangle(15, 15, 80, 80, { roughness: 0.5, fill: 'red' });
rc.rectangle(120, 15, 80, 80, { roughness: 2.8, fill: 'blue' });
rc.rectangle(220, 15, 80, 80, { bowing: 6, stroke: 'green', strokeWidth: 3 });
```

### SVG Paths

![Rough.js paths](https://roughjs.com/images/m5.png)

```js
rc.path('M80 80 A 45 45, 0, 0, 0, 125 125 L 125 80 Z', { fill: 'green' });
rc.path('M230 80 A 45 45, 0, 1, 0, 275 125 L 275 80 Z', { fill: 'purple' });
rc.path('M80 230 A 45 45, 0, 0, 1, 125 275 L 125 230 Z', { fill: 'red' });
rc.path('M230 230 A 45 45, 0, 1, 1, 275 275 L 275 230 Z', { fill: 'blue' });
```

SVG Path with simplification:

![Rough.js texas map](https://roughjs.com/images/m9.png) ![Rough.js texas map](https://roughjs.com/images/m10.png)

## Examples

![Rough.js US map](https://roughjs.com/images/m6.png)

[View examples here](https://github.com/pshihn/rough/wiki/Examples)

## API & Documentation

[Full Rough.js API](https://github.com/pshihn/rough/wiki)

## Credits

Some of the core algorithms were adapted from [handy](https://www.gicentre.net/software/#/handy/) processing lib.

Algorithm to convert SVG arcs to Canvas [described here](https://www.w3.org/TR/SVG/implnote.html) was adapted from [Mozilla codebase](https://hg.mozilla.org/mozilla-central/file/17156fbebbc8/content/svg/content/src/nsSVGPathDataParser.cpp#l887)

## Contributors

### Financial Contributors

Become a financial contributor and help us sustain our community. [[Contribute](https://opencollective.com/rough/contribute)]

#### Individuals

<a href="https://opencollective.com/rough"><img src="https://opencollective.com/rough/individuals.svg?width=890"></a>

#### Organizations

Support this project with your organization. Your logo will show up here with a link to your website. [[Contribute](https://opencollective.com/rough/contribute)]

<a href="https://www.diagrams.net/"><img src="https://avatars.githubusercontent.com/u/1769238?s=64&v=4"></a>
<a href="https://terrastruct.com/"><img width="64" height="64" src="https://roughjs.com/images/sponsors/terrastruct.png"></a>
<a href="https://opencollective.com/rough/organization/0/website"><img src="https://opencollective.com/rough/organization/0/avatar.svg"></a>
<a href="https://opencollective.com/rough/organization/1/website"><img src="https://opencollective.com/rough/organization/1/avatar.svg"></a>
<a href="https://opencollective.com/rough/organization/2/website"><img src="https://opencollective.com/rough/organization/2/avatar.svg"></a>
<a href="https://opencollective.com/rough/organization/3/website"><img src="https://opencollective.com/rough/organization/3/avatar.svg"></a>
<a href="https://opencollective.com/rough/organization/4/website"><img src="https://opencollective.com/rough/organization/4/avatar.svg"></a>
<a href="https://opencollective.com/rough/organization/5/website"><img src="https://opencollective.com/rough/organization/5/avatar.svg"></a>
<a href="https://opencollective.com/rough/organization/6/website"><img src="https://opencollective.com/rough/organization/6/avatar.svg"></a>
<a href="https://opencollective.com/rough/organization/7/website"><img src="https://opencollective.com/rough/organization/7/avatar.svg"></a>
<a href="https://opencollective.com/rough/organization/8/website"><img src="https://opencollective.com/rough/organization/8/avatar.svg"></a>
<a href="https://opencollective.com/rough/organization/9/website"><img src="https://opencollective.com/rough/organization/9/avatar.svg"></a>

## License
[MIT License](https://github.com/pshihn/rough/blob/master/LICENSE) (c) [Preet Shihn](https://twitter.com/preetster)


CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file.

# [4.5.0] - 2021-05-09
* Better algorithm for nested and intersecting paths https://github.com/rough-stuff/rough/issues/183
* Fixed "dots" fill when roughness was <1 Itw as creatings weird shapes https://github.com/rough-stuff/rough/issues/193
* Configure precision when rendering to canvas as well as SVG using `fixedDecimalPlaceDigits` property.
* Solid fill was broken for Arcs if arc angle was > 180 degrees
* Remove notch from ellipses when roughness = 0


# [4.4.0] - 2021-05-09

* Added `preserveVertices` option when drawing shapes. Especially useful in paths. When rendering a shape, the vertices or the end points of the shape are not randomized if this is set to TRUE. This allows connected segments to always be connected. 

## [4.3.0] - 2020-05-11

* Added options to draw dashed lines - *strokeLineDash, strokeLineDashOffset, fillLineDash, fillLineDashOffset*
* Added option to disable double stroking effect - *disableMultiStroke, disableMultiStrokeFill*.
* Bug fixes to solid fill in SVG which was not obeying evenodd rules by default

## [4.1.0] - 2020-01-13

* Added ability to **fill** non-svg curves

## [4.0.0] - 2020-01-13

* Add optional seeding for randomness to ensure shapes generated with same arguments result in same vectors
* Implemented a new algorithm for hachure generation based on scanlines. Smaller in code size, and about 20% faster
* Algorithm update - adjust shape randomness and curve-step-counts based on the size of the shape
* Removed async/worker builds - can be achieved in the app level, so no need to be in the lib
* Support no-stroke sketching. `stroke: "none"` will not generate outline vectors anymore
* Removed `sunburst` fill style - it had a lot of corner cases where it did not work, and not very popular.

## [3.1.0] - 2019-03-14

* Added three new fill styles: **sunburst**, **dashed**, and **zigzag-line**
* Added three new properties in *Options* to support these fill styles:
* **dashOffset** - length of dashes in dashed fill
* **dashGap** - length of gap between dashes in dashed fill
* **zigzagOffset** - width of zigzag triangle when using zigzag-lines fill





src/webgl.ts
import { ProgramInfo, createProgramInfo, resizeCanvasToDisplaySize, setUniforms, AttributeValues, setAttributes } from './webgl-utils';
import { Point, BezCurve, estBezLength } from './geometry';
import { Drawable, Op } from './core';
import { DEFAULT_OPTIONS } from './generator';
import { parseColor } from './colors';
import { vsSource, fsSource } from './webgl-shaders';

interface AttributeInfo {
  name: string;
  numComponents: number;
  offset: number;
}
const FRAME_SIZE = 16;
const STRIDE = FRAME_SIZE * 4;

const ATTRIBUTES: AttributeInfo[] = [
  { name: 'a_p1', numComponents: 2, offset: 0 },
  { name: 'a_cp1', numComponents: 2, offset: 2 },
  { name: 'a_cp2', numComponents: 2, offset: 4 },
  { name: 'a_p2', numComponents: 2, offset: 6 },
  { name: 'a_color', numComponents: 4, offset: 8 },
  { name: 'a_thickness', numComponents: 1, offset: 12 },
  { name: 'a_numVerts', numComponents: 1, offset: 13 },
  { name: 'a_vertexId', numComponents: 2, offset: 14 }
];

export class RoughWebGL {
  private canvas: HTMLCanvasElement;
  private gl: WebGLRenderingContext;
  private pi: ProgramInfo;
  private defaultOptions = JSON.parse(JSON.stringify(DEFAULT_OPTIONS));

  constructor(canvas: HTMLCanvasElement) {
    this.canvas = canvas;
    this.gl = canvas.getContext('webgl')!;
    if (!this.gl) {
      throw new Error('Unable to initialize WebGL. Your browser or machine may not support it.');
    }
    this.pi = createProgramInfo(this.gl, [vsSource, fsSource]);
  }

  draw(drawings: Drawable | Drawable[], clearScene = false) {
    const drawables = Array.isArray(drawings) ? drawings : [drawings];
    if (clearScene) {
      this.clearCanvas();
    }
    this.initializeCanvas();

    for (const drawable of drawables) {
      const sets = drawable.sets || [];
      const o = drawable.options || this.defaultOptions;
      for (const drawing of sets) {
        switch (drawing.type) {
          case 'path':
            // const curvePoints = this.extractCurvePoints(drawing.ops);
            let curves: BezCurve[] = [
              [
                [10, 30],
                [120, 160],
                [180, 10],
                [220, 140]
              ],
              [
                [10, 80],
                [120, 210],
                [180, 60],
                [220, 190]
              ],
              [
                [10, 110],
                [120, 240],
                [180, 90],
                [220, 220]
              ]
            ];
            curves = [...curves, ...curves];
            curves = [...curves, ...curves];
            curves = [...curves, ...curves];
            curves = [...curves, ...curves];
            curves = [...curves, ...curves];
            curves = [...curves, ...curves];
            curves = [...curves, ...curves];
            curves = [...curves, ...curves];

            const aa1 = performance.now();
            this.drawCurves(curves, o.strokeWidth, o.stroke === 'none' ? 'transparent' : o.stroke);
            console.log('draw curves', performance.now() - aa1);
            break;
          case 'fillSketch': {
            let fweight = o.fillWeight;
            if (fweight < 0) {
              fweight = o.strokeWidth / 2;
            }
            const curvePoints = this.extractCurvePoints(drawing.ops);
            this.drawCurves(curvePoints, fweight, o.fill || '');
            break;
          }
        }
      }
    }
  }

  private extractCurvePoints(ops: Op[]): BezCurve[] {
    const curves: BezCurve[] = [];
    let lastPoint: Point = [0, 0];
    for (const op of ops) {
      switch (op.op) {
        case 'move':
          lastPoint = op.data as Point;
          break;
        case 'bcurveTo':
          const curve: BezCurve = [
            lastPoint,
            [op.data[0], op.data[1]],
            [op.data[2], op.data[3]],
            [op.data[4], op.data[5]]
          ];
          curves.push(curve);
          lastPoint = curve[3];
          break;
      }
    }
    return curves;
  }

  clearCanvas() {
    this.gl.clearColor(0, 0, 0, 0);
    this.gl.clear(this.gl.COLOR_BUFFER_BIT);
  }

  private initializeCanvas() {
    // TODO: need this?
    resizeCanvasToDisplaySize(this.canvas);
    this.gl.viewport(0, 0, this.gl.canvas.width, this.gl.canvas.height);
  }

  private drawCurves(curves: BezCurve[], lineWidth: number, color: string) {
    console.log('curve count', curves.length);

    // Initialize buffer data
    const tt1 = performance.now();
    const parsedColor = parseColor(color);
    const thickness = lineWidth / (window.devicePixelRatio || 1);
    const data: number[] = [];
    let curveIndex = 0;
    for (const curve of curves) {
      // validate curve
      const numVerts = Math.ceil(estBezLength(curve) / 2);
      if (numVerts < 3) {
        continue;
      }
      for (let i = 0; i < numVerts; i++) {
        curve.forEach((p) => data.push(...p)); // 8
        data.push(
          ...parsedColor, // 4
          thickness,      // 1
          numVerts,       // 1
          i, -1           // 2
        );
        curve.forEach((p) => data.push(...p)); // 8
        data.push(
          ...parsedColor, // 4
          thickness,      // 1
          numVerts,       // 1
          i, 1           // 2
        );
      }
      curveIndex++;
    }
    const tt2 = performance.now();

    // Initialize buffer
    const attributeBuffer = this.gl.createBuffer()!;
    const bufferType = this.gl.ARRAY_BUFFER;
    this.gl.bindBuffer(bufferType, attributeBuffer);
    this.gl.bufferData(bufferType, new Float32Array(data), this.gl.DYNAMIC_DRAW);

    const tt3 = performance.now();

    // use program
    this.gl.useProgram(this.pi.program);

    // set attributes
    const attrValues: AttributeValues = {};
    ATTRIBUTES.forEach((attr) => {
      attrValues[attr.name] = {
        buffer: attributeBuffer,
        numComponents: attr.numComponents,
        offset: attr.offset * 4,
        stride: STRIDE
      };
    });
    setAttributes(this.pi.attribSetters, attrValues);

    const tt4 = performance.now();

    // set uniforms
    setUniforms(this.pi.uniformSetters, {
      u_resolution: [this.gl.canvas.width, this.gl.canvas.height]
    });

    // Draw
    const tt5 = performance.now();
    this.gl.drawArrays(this.gl.TRIANGLE_STRIP, 0, data.length / FRAME_SIZE);
    const tt6 = performance.now();

    console.log('init data', tt2 - tt1);
    console.log('init buffer', tt3 - tt2);
    console.log('set attributes', tt4 - tt3);
    console.log('set uniform', tt5 - tt4);
    console.log('draw', tt6 - tt5);
    console.log('data lengths', data.length, data.length / FRAME_SIZE);
  }
}